ft2demos-2.5.5/src/ftlint.1
.TH FTLINT 1 "December 2014" "Freetype 2.5.5"
.
.
.SH NAME
.
ftlint \- simple font tester
.
.
.SH SYNOPSIS
.
.B ftlint
.I ppem
.IR fontname [ .ttf | .ttc "] .\|.\|."
.
.
.SH DESCRIPTION
.
.B ftlint
opens the given font(s), loads and renders all glyphs at the given ppem
value, and reports failing glyphs if there are less than ten fails, or the
number of fails only if there are more fails.
.
.PP
This program does not have any options.
It is part of the FreeType demos package.
.
.\" eof


ft2demos-2.5.5/ChangeLog
2014-12-30  Werner Lemberg  <wl@gnu.org>

	* Version 2.5.5 released.
	=========================


	Tag sources with `VER-2-5-5'.

	* README: Updated.
	* src/*.1: Updated.

	* Makefile (dist): Fix typos.

2014-12-11  Werner Lemberg  <wl@gnu.org>

	* Makefile (dist): Use older POSIX standard for `tar'.

	Apparently, BSD tar isn't capable yet of handling POSIX-1.2001
	(contrary to GNU tar), so force the POSIX-1.1988 format.

	Problem reported by Stephen Fisher <sfisher@SDF.ORG>.

2014-12-06  Werner Lemberg  <wl@gnu.org>

	* Version 2.5.4 released.
	=========================


	Tag sources with `VER-2-5-4'.

	* README: Updated.
	* src/*.1: Updated.

2014-11-22  Werner Lemberg  <wl@gnu.org>

	Show FreeType version in demo program help screens.

	* src/ftdiff.c (event_help), src/ftgrid.c (event_help),
	src/ftmulti.c (Help), src/ftstring.c (event_help), src/ftview.c
	(event_help): Implement it.

2014-05-11  Werner Lemberg  <wl@gnu.org>

	Fix Savannah bug #42182.

	* src/ttdebug.c (main): Fix typos.

2014-03-06  Werner Lemberg  <wl@gnu.org>

	* Version 2.5.3 released.
	=========================


	Tag sources with `VER-2-5-3'.

	* README: Updated.
	* src/*.1: Updated.

2014-03-04  Werner Lemberg  <wl@gnu.org>

	* Makefile (LINK_LIBS) [unixdev]: Don't use pkg-config for bzip2.

	Not all platforms have `bzip2.pc'.  We also add `-lz' just for
	completeness.

2014-01-30  Werner Lemberg  <wl@gnu.org>

	[ftbench] Allow `-s 0' for better testing of FT_LOAD_NO_SCALE.

	* src/ftbench.c (main) <'s'>: Don't alter value 0.
	Don't call `FT_Set_Pixel_Sizes' for size == 0.
	Don't call tests for size == 0 that need valid scaling factors.
	(usage): Updated.
	(benchmark): Return meaningful timer value for error-only calls.
	(test_embolden): Enable test.  I think there is no longer a reason
	to not run it.

	* src/ftbench.1: Updated.

2014-01-28  Werner Lemberg  <wl@gnu.org>

	* src/ftgrid.c (grid_status_rescale_initial): Avoid NaN values.
	Remove forgotten debug message.

2014-01-28  Werner Lemberg  <wl@gnu.org>

	* graph/grblit.c (grBlitGlyphToBitmap): Don't crash on empty glyphs.
	Problem reported by Dave Arnold <darnold@adobe.com>.

2014-01-03  Werner Lemberg  <wl@gnu.org>

	[ftgrid] Show (some) blue zones.

	* src/ftgrid.c (af_glyph_hints_get_segment_offset): Update
	prototype.
	(GridStatucRec): New member `blue_color'.
	(grid_status_display): Change `on' color to red.
	Set `blue' color.
	(grid_hint_draw_segment): Show associated blue zones.
	(event_help): Updated.

2014-01-03  Werner Lemberg  <wl@gnu.org>

	[ftgrid] Minor.

	* src/ftgrid.c (GridStatusRec): Rename `conic_color' to `off_color'.
	Remove unused `cubic_color' member.
	(grid_status_display, grid_status_draw_outline): Updated.

2013-12-22  Werner Lemberg  <wl@gnu.org>

	* Makefile (LINK_LIBS) [unixdev]: Use pkg-config to get link flags.

	This corresponds to a similar change in current FreeType to link
	with HarfBuzz in development mode.

2013-12-08  Werner Lemberg  <wl@gnu.org>

	* Version 2.5.2 released.
	=========================


	Tag sources with `VER-2-5-2'.

	* README: Updated.
	* src/*.1: Updated.

2013-11-27  Werner Lemberg  <wl@gnu.org>

	[ftgrid] Fix compilation in non-debug mode.

	* src/ftgrid.c: Guard autofit dump functions with FT_DEBUG_AUTOFIT
	clause.
	Update prototypes.
	(grid_hint_draw_segment): Guard with FT_DEBUG_AUTOFIT clause.
	(grid_status_draw_outline): Guard call to `grid_hint_draw_segment'
	with FT_DEBUG_AUTOFIT clause.
	(Process_Event) <'1'>, <'2'>, <'3'>: Don't mess around with trace
	levels, which should always stay internal to the FreeType library.
	Update calls to dump functions.

2013-11-25  Werner Lemberg  <wl@gnu.org>

	* Version 2.5.1 released.
	=========================


	Tag sources with `VER-2-5-1'.

	* README: Updated.
	* src/*.1: Updated.

2013-11-21  Werner Lemberg  <wl@gnu.org>

	[ftgrid] Make keys `1', `2', and `3' work again.

	* src/ftgrid.c (Process_Event) <'1'>, <'2'>, <'3'>: Set
	`trace_afhints'	trace level to value 7 temporarily since the autofit
	dumping functions use `FT_TRACE7' since 2012-07-06.

2013-11-21  Werner Lemberg  <wl@gnu.org>

	[ftgrid] Fix compilation with C++.

	* src/ftgrid.c (_af_debug_*): Add `extern' keyword.

2013-11-20  Werner Lemberg  <wl@gnu.org>

	Increase file path size.

	* src/compos.c (main), src/ftchkwd.c (main), src/ftdump.c (main),
	src/ftlint.c (main), src/ftmemchk.c (main), src/ftsbit.c (main),
	src/fttimer.c (main), src/fttry.c (main): Increase array size of
	`filename' and `alt_filename' to 1024.

2013-11-13  Werner Lemberg  <wl@gnu.org>

	* Makefile (FT_H): Update to new header layout of FreeType.

2013-11-05  Werner Lemberg  <wl@gnu.org>

	[ttdebug] Improve display of point tags.

	* src/ttdebug.c (old_tag_to_new): Removed, no longer used.
	(display_changed_points, show_points_table): Display tags with
	mnemonic characters instead of ORed numbers.
	(RunIns) <'h'>: Updated.

2013-11-05  Werner Lemberg  <wl@gnu.org>

	[ttdebug] Enable debugging of composite glyphs (again).

	This was lost while improving key handling.

	* src/ttdebug.c (RunIns): Introduce variable `really_leave' to
	insert one stop before really leaving the current glyph range.
	Without that, inspecting the current data right after finishing the
	hinting of a subglyph wouldn't be possible.
	(main): Stay in loop even if `FT_Load_Glyph' returns zero.

2013-11-04  Werner Lemberg  <wl@gnu.org>

	[ttdebug] Provide man page and compile program by default.

	* Makefile (EXES): Add `ttdebug'.

	* src/ttdebug.1: New file.

2013-11-01  Werner Lemberg  <wl@gnu.org>

	[ttdebug] Make `B' emit a function backtrace.

	This needs the today's version of the FreeType library.

	* src/ttdebug (RunIns) <'h'>: Updated.
	<'B'>: Print backtrace.

2013-11-01  Werner Lemberg  <wl@gnu.org>

	[ttdebug] Make `p' set a breakpoint at previous position.

	* src/ttdebug.c (RunIns): Store information about last position in
	`last_IP' and `last_range'.
	<'h'>, <'c'>, <'f'>, <'n'>, <'s'>: Updated.

ft2demos-2.5.5/src/ftlint.c
/****************************************************************************/
/*                                                                          */
/*  The FreeType project -- a free and portable quality font engine         */
/*                                                                          */
/*  Copyright 1996-1998, 2001, 2013 by                                      */
/*  D. Turner, R.Wilhelm, and W. Lemberg                                    */
/*                                                                          */
/*  ftlint: a simple font tester. This program tries to load all the        */
/*          glyphs of a given font.                                         */
/*                                                                          */
/*  NOTE:  This is just a test program that is used to show off and         */
/*         debug the current engine.                                        */
/*                                                                          */
/****************************************************************************/

#include <ft2build.h>
#include FT_FREETYPE_H

#include <stdio.h>
#include <stdlib.h>
#include <string.h>


#define  xxTEST_PSNAMES

#define gettext( x )  ( x )

  FT_Error      error;

  FT_Library    library;
  FT_Face       face;
  FT_Size       size;
  FT_GlyphSlot  slot;

  unsigned int  num_glyphs;
  int           ptsize;

  int  Fail;
  int  Num;



  static void  Usage( char*  name )
  {
    printf( "ftlint: simple font tester -- part of the FreeType project\n" );
    printf( "----------------------------------------------------------\n" );
    printf( "\n" );
    printf( "Usage: %s ppem fontname[.ttf|.ttc] [fontname2..]\n", name );
    printf( "\n" );

    exit( 1 );
  }


  static void  Panic( const char*  message )
  {
    fprintf( stderr, "%s\n  error code = 0x%04x\n", message, error );
    exit(1);
  }


  int  main( int  argc, char**  argv )
  {
    int           i, file_index;
    unsigned int  id;
    char          filename[1024 + 4];
    char          alt_filename[1024 + 4];
    char*         execname;
    char*         fname;


    execname = argv[0];

    if ( argc < 3 )
      Usage( execname );

    if ( sscanf( argv[1], "%d", &ptsize ) != 1 )
      Usage( execname );

    error = FT_Init_FreeType( &library );
    if (error) Panic( "Could not create library object" );

    /* Now check all files */
    for ( file_index = 2; file_index < argc; file_index++ )
    {
      fname = argv[file_index];

      /* try to open the file with no extra extension first */
      error = FT_New_Face( library, fname, 0, &face );
      if (!error)
      {
        printf( "%s: ", fname );
        goto Success;
      }


      if ( error == FT_Err_Unknown_File_Format )
      {
        printf( "unknown format\n" );
        continue;
      }

      /* ok, we could not load the file, try to add an extension to */
      /* its name if possible..                                     */

      i     = strlen( fname );
      while ( i > 0 && fname[i] != '\\' && fname[i] != '/' )
      {
        if ( fname[i] == '.' )
          i = 0;
        i--;
      }

      filename[1024] = '\0';
      alt_filename[1024] = '\0';

      strncpy( filename, fname, 1024 );
      strncpy( alt_filename, fname, 1024 );

#ifndef macintosh
      if ( i >= 0 )
      {
        strncpy( filename + strlen( filename ), ".ttf", 4 );
        strncpy( alt_filename + strlen( alt_filename ), ".ttc", 4 );
      }
#endif
      i     = strlen( filename );
      fname = filename;

      while ( i >= 0 )
#ifndef macintosh
        if ( filename[i] == '/' || filename[i] == '\\' )
#else
        if ( filename[i] == ':' )
#endif
        {
          fname = filename + i + 1;
          i = -1;
        }
        else
          i--;

      printf( "%s: ", fname );

      /* Load face */
      error = FT_New_Face( library, filename, 0, &face );
      if (error)
      {
        if (error == FT_Err_Unknown_File_Format)
          printf( "unknown format\n" );
        else
          printf( "could not find/open file (error: %d)\n", error );
        continue;
      }
      if (error) Panic( "Could not open file" );

  Success:
      num_glyphs = face->num_glyphs;

#ifdef  TEST_PSNAMES
      {
        const char*  ps_name = FT_Get_Postscript_Name( face );

        printf( "[%s] ", ps_name ? ps_name : "." );
      }
#endif

      error = FT_Set_Char_Size( face, ptsize << 6, ptsize << 6, 72, 72 );
      if (error) Panic( "Could not set character size" );

      Fail = 0;
      {
        for ( id = 0; id < num_glyphs; id++ )
        {
          error = FT_Load_Glyph( face, id, FT_LOAD_DEFAULT );
          if (error)
          {
            if ( Fail < 10 )
              printf( "glyph %4u: 0x%04x\n" , id, error );
            Fail++;
          }
        }
      }

      if ( Fail == 0 )
        printf( "OK.\n" );
      else
        if ( Fail == 1 )
          printf( "1 fail.\n" );
        else
          printf( "%d fails.\n", Fail );

      FT_Done_Face( face );
    }

    FT_Done_FreeType(library);
    exit( 0 );      /* for safety reasons */

    return 0;       /* never reached */
  }


/* End */
